lib/utils/display.js

```diff
@@ -0,0 +1,124 @@
+const { inspect } = require('util')
+const npmlog = require('npmlog')
+const log = require('./log-shim.js')
+const { explain } = require('./explain-eresolve.js')
+
+const _logHandler = Symbol('logHandler')
+const _eresolveWarn = Symbol('eresolveWarn')
+const _log = Symbol('log')
+const _npmlog = Symbol('npmlog')
+
+class Display {
+  constructor () {
+    // pause by default until config is loaded
+    this.on()
+    log.pause()
+  }
+
+  on () {
+    process.on('log', this[_logHandler])
+  }
+
+  off () {
+    process.off('log', this[_logHandler])
+    // Unbalanced calls to enable/disable progress
+    // will leave change listeners on the tracker
+    // This pretty much only happens in tests but
+    // this removes the event emitter listener warnings
+    log.tracker.removeAllListeners()
+  }
+
+  load (config) {
+    const {
+      color,
+      timing,
+      loglevel,
+      unicode,
+      progress,
+      heading = 'npm',
+    } = config
+
+    // XXX: decouple timing from loglevel
+    if (timing && loglevel === 'notice') {
+      log.level = 'timing'
+    } else {
+      log.level = loglevel
+    }
+
+    log.heading = heading
+
+    if (color) {
+      log.enableColor()
+    } else {
+      log.disableColor()
+    }
+
+    if (unicode) {
+      log.enableUnicode()
+    } else {
+      log.disableUnicode()
+    }
+
+    // Progress is on stderr so see if term supports that
+    const stderrTTY = process.stderr.isTTY
+    const dumbTerm = process.env.TERM === 'dumb'
+    const stderrNotDumb = stderrTTY && !dumbTerm
+    // if it's more than error, don't show progress
+    const quiet = log.levels[log.level] > log.levels.error
+
+    if (progress && stderrNotDumb && !quiet) {
+      log.enableProgress()
+    } else {
+      log.disableProgress()
+    }
+
+    // Resume displaying logs now that we have config
+    log.resume()
+  }
+
+  log (...args) {
+    this[_logHandler](...args)
+  }
+
+  [_logHandler] = (level, ...args) => {
+    try {
+      this[_log](level, ...args)
+    } catch (ex) {
+      try {
+        // if it crashed once, it might again!
+        this[_npmlog]('verbose', `attempt to log ${inspect(args)} crashed`, ex)
```

Comment:
This feels like extremely legacy overly-optimized behavior and I wonder if we need this still